### PR TITLE
Add integration tests

### PR DESCRIPTION
**Pull request by DangerOnTheRanger**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 1]
```diff
@@ -0,0 +1,918 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+
+	v1 "k8s.io/api/core/v1"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+)
+
+func makePolicy(name string) *admissionregistrationv1alpha1.ValidatingAdmissionPolicy {
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: Can you move all these helper functions to the bottom of the file before the tests? Just makes it easier to read when the tests come first.

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 2]
```diff
@@ -0,0 +1,918 @@
...
+				t.Logf("actual error: %v", err)
+				t.Logf("expected error: %v", testcase.err)
+				t.Fatal("got error but expected none")
+			}
+
+			if err.Error() != testcase.err {
+				t.Logf("actual validation error: %v", err)
+				t.Logf("expected validation error: %v", testcase.err)
+				t.Error("unexpected validation error")
+			}
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:476]: Thoughts on adding the status reason check now? 
```
reason := err.(apierrors.APIStatus).Status().Reason
if reason != testcase.failureReason {
    t.Logf("actual error reason: %v", reason)
    t.Logf("expected failure reason: %v", testcase.failureReason)
    t.Error("unexpected error reason")
}
```

Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go:476]: Right, I'll mark this as resolved when I add it.

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 1 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: Fixed!

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+			},
+			namespace: &v1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-k8s",
+				},
+			},
+			err: "",
+		},
+		{
+			name: "namespace name does NOT contain suffix enforced by validating admission policyusing, object metadata fields",
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:109]: Let's add a version of this test, but using Forbidden as the expected status reason

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+			if reason != testcase.failureReason {
+				t.Logf("actual error reason: %v", reason)
+				t.Logf("expected failure reason: %v", testcase.failureReason)
+				t.Error("unexpected error reason")
+			}
+		})
+	}
+}
+
+func TestNoResource(t *testing.T) {
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: Can we use a better name for this tests?

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
+	policy.Spec.Validations = []admissionregistrationv1alpha1.Validation{
+		{
+			Expression: "true",
+		},
+	}
+	if _, err := client.AdmissionregistrationV1alpha1().ValidatingAdmissionPolicies().Create(context.TODO(), policy, metav1.CreateOptions{}); err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO (DangerOnTheRanger): add and subsequently remove config map here (instead of policy being invalid from the start)
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: Is this TODO still relevant?

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 4 above)
Comment (andrewsykim) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: `Test_ValidatingAdmissionPolicy_ParamObjectDeleted`?

Comment (andrewsykim) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: Seems like the test needs to be updated to remove the ConfigMap though, do you want to just remove this test for now and add it as a follow-up?

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 6]
```diff
@@ -0,0 +1,937 @@
...
+	autopassingSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "autopassing-secret",
+			Labels: map[string]string{
+				"paramIdent":             "someVal",
+				"autopass-binding-label": "true",
+			},
+		},
+	}
+	if _, err := client.CoreV1().Secrets("default").Create(context.TODO(), autopassingSecret, metav1.CreateOptions{}); err != nil {
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:793]: Similar to the other tests, we should validate the actual error and the status reason

Comment (andrewsykim) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go:793]: (might be better to remove this test for now and add it back as a follow-up once we have the error check, up to you)

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 5 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: The test as it stands never makes the config map - implementing the TODO would temporarily put the policy binding in a valid state, and there might be separate codepaths that get tested if that happens; do we test reconciling with a deleted param object anywhere else?

I think if we have a similar reconciliation test elsewhere I'd be on board for removing the TODO for sure, I'm a little worried there could be untested code otherwise.

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 4 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: Yeah, I think we can remove the test from this PR for now - I'll mark this as resolved when I do that.

At the moment I'm still in favor of the test getting implemented with [this TODO](https://github.com/cici37/kubernetes/pull/9#discussion_r1014375606) though that's a separate concern I think.

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 6 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go:793]: Right, I'll mark this as resolved once I add that.

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 5 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go]: (marking this as resolved since the test is removed for now)

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 3 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go:109]: Fixed!

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 7]
```diff
@@ -0,0 +1,942 @@
...
+
+func makeConfigParams(name string, data map[string]string) *v1.ConfigMap {
+	return &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+		Data:       data,
+	}
+}
+
+func checkForFailedRule(t *testing.T, err error) {
+	if !strings.Contains(err.Error(), "failed expression") {
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:927]: For a follow-up: given it's possible to know the expected error string, seems like we should just compare the full string

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 8]
```diff
@@ -0,0 +1,942 @@
...
+	if _, err := client.CoreV1().Secrets("default").Create(context.TODO(), condpassingSecret, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("expected secret creation to succeed, got: %s", err)
+	}
+
+	condfailingSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "condfailing-secret",
+			Labels: map[string]string{
+				"paramIdent":             "someVal",
+				"condpass-binding-label": "true",
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:815]: Is `"condpass-binding-label": "true"` expected for `condfailingSecret`?

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 9]
```diff
@@ -0,0 +1,942 @@
...
+	if _, err := client.CoreV1().Secrets("default").Create(context.TODO(), autopassingSecret, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("expected secret creation to succeed, got: %s", err)
+	}
+
+	condpassingSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "prefix-condpassing-secret",
+			Labels: map[string]string{
+				"paramIdent":             "someVal",
+				"condpass-binding-label": "true",
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:802]: I think I'm just missing something, but where is the expression checking for the `condpass-binding-label`? 
```
Expression: "params.data.autofail != 'true' && (params.data.conditional == 'false' || object.metadata.name.startsWith(params.data.check))",
```

Comment (andrewsykim) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go:802]: Oh, this is passing because of the name prefix, not the label and `condpass-binding-label` is for the object selector

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go [hunk 10]
```diff
@@ -0,0 +1,942 @@
...
+		t.Fatal(err)
+	}
+	autopassBinding := withBindingExistsLabels([]string{"autopass-binding-label"}, policy, makeBinding("autopass-binding", "test-policy", "autopass-params"))
+	if _, err := client.AdmissionregistrationV1alpha1().ValidatingAdmissionPolicyBindings().Create(context.TODO(), autopassBinding, metav1.CreateOptions{}); err != nil {
+		t.Fatal(err)
+	}
+
+	condpassParams := makeConfigParams("condpass-params", map[string]string{
+		"autofail":    "false",
+		"conditional": "true",
```
Comment (andrewsykim) [inline test/integration/apiserver/cel/validatingadmissionpolicy_test.go:754]: I find the combination of these params a bit hard to follow -- wondering if we can make the pass/fail conditions for mulitple bindings more obvious :thinking:

**Review by andrewsykim — COMMENTED** [review]
LGTM

Remaining can be addressed in follow-up

File: test/integration/apiserver/cel/validatingadmissionpolicy_test.go (hunk 10 above)
Comment (DangerOnTheRanger) [reply test/integration/apiserver/cel/validatingadmissionpolicy_test.go:754]: Would renaming them help? I'm not sure how much there is to simplify with the logic behind the expression itself (one is directly checked by the expression, the other being a switch that enables another branch of the expression that uses param data).
